src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,272 @@
+                    process.StartInfo = new ProcessStartInfo
+                    {
+                        Arguments = GetCommandLineArguments(MSBuildBinPath),
+                        CreateNoWindow = true,
+                        FileName = GetProcessFileName(ProcessFileName),
+                        RedirectStandardInput = true,
+                        RedirectStandardOutput = true,
+                        UseShellExecute = false,
+                        WorkingDirectory = Environment.CurrentDirectory,
+                    };
+
+                    // Place the output in the queue which handles logging messages coming through on StdOut
+                    process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
+
+                    process.Exited += (sender, args) => semaphore.Release();
+
+                    try
+                    {
+                        Log.LogMessage(MessageImportance.Low, "Running command: \"{0}\" {1}", process.StartInfo.FileName, process.StartInfo.Arguments);
+
+                        process.Start();
+
+                        WriteArguments(process.StandardInput);
+
+                        process.BeginOutputReadLine();
+
+                        semaphore.Wait(_cancellationTokenSource.Token);
+
+                        if (!process.HasExited)
+                        {
+                            try
+                            {
+                                process.Kill();
+                            }
+                            catch (InvalidOperationException)
+                            {
+                                // The process may have exited, in this case ignore the exception
+                            }
+                        }
+                    }
+                    catch (Exception e) when (
+                        e is OperationCanceledException
+                        || (e is AggregateException aggregateException && aggregateException.InnerException is OperationCanceledException))
+                    {
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.LogErrorFromException(e);
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        /// <summary>
+        /// Gets the command-line arguments to use when launching the process that executes the restore.
+        /// </summary>
+        internal string GetCommandLineArguments(string msbuildBinPath)
+        {
+            return string.Concat(
+#if IS_CORECLR
+                "\"", Path.Combine(ThisAssemblyLazy.Value.DirectoryName, Path.ChangeExtension(ThisAssemblyLazy.Value.Name, ".Console.dll")), "\"",
+                " ", "\"", Path.Combine(msbuildBinPath, "MSBuild.dll"), "\"",
+#else
+                "\"", Path.Combine(msbuildBinPath, "MSBuild.exe"), "\"",
+#endif
+                " ", "\"", IsSolutionPathDefined ? SolutionPath : ProjectFullPath, "\"");
```

Comment:
minor nitpick/suggestion

```suggestion
                " \"", IsSolutionPathDefined ? SolutionPath : ProjectFullPath, "\"");
```

Reply:
I had typed it up like this originally but thought it wasn't obvious enough that the command-line arguments are wrapped in quotes and space delimited.  Do you know if there's a performance penalty with more arguments?  Or do you just prefer this syntax?

Reply:
I expect if a perf measurement was made, that it would be smaller than measurement jitter. While the number of chars copied in `string.Concat` is the same, an additional string needs to be pushed onto the stack, so that's at least additional 1 CPU instruction. And in order to be argumentative, we could also talk about very theoretical issues like CPU cache hits/misses, but I bet we couldn't get consistent benchmarks to measure anything meaningful.

Honestly, I'm not sure I really have a preference. I created the suggestions as a gut reaction, but can see how some people might prefer the two single character strings instead. Feel free to leave it as is.